### DISCUSSION TITLE
`derives` doesn't compile on higher kinded types

### DISCUSSION THREAD

**Discussion by kubukoz** [scala/scala3#17875] [Feature Requests]
## Compiler version

3.0.0-M3 - circe isn't publishing for RC1 yet, and that's the library I found this with.

## Minimized code

https://github.com/kubukoz/derives-repro

```Scala
type Id[A] = A

trait Codec[A]
object Codec:
  def derived[A]: Codec[A] = ???


enum Expr[F[_]] derives Codec:
  case Term(name: F[String])
  case Apply(on: Expr[F], param: Expr[F])

//this works:
// object Expr:
//   given Codec[Expr[Id]] = Codec.derived
```

## Output

```scala
Compiling root (1 Scala source)
[E] [E1] src/main/scala/Main.scala:8:25
[E]      Expr cannot be unified with the type argument of Codec
[E]      L8: enum Expr[F[_]] derives Codec:
[E]                                  ^
[E] src/main/scala/Main.scala: L8 [E1]
```

## Expectation

Code compiles and something roughly like this is generated in the companion:

```scala
object Expr:
  given [F[_]]: Codec[Expr[F]] = Codec.derived
```

**Comment by smarter** [comment]
Needs a minimization without dependencies.

**Comment by kubukoz** [comment]
Done ;)

**Comment by smarter** [comment]
Here's the logic which decides how to deal with `derives`: https://github.com/lampepfl/dotty/blob/5eb32588aef946e8c74f72bde6a7d26684fa08e1/compiler/src/dotty/tools/dotc/typer/Deriving.scala#L63-L84
And the exact rules are given just below: https://github.com/lampepfl/dotty/blob/5eb32588aef946e8c74f72bde6a7d26684fa08e1/compiler/src/dotty/tools/dotc/typer/Deriving.scala#L108-L153

Can these rules be generalized to support your usecase?

**Comment by smarter** [comment]
Moved to the feature request repo since it seems like the current implementation is working as expected, just that some users have greater expectations :).
